### PR TITLE
Misc performance optimizations

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
getQueryDetails.api performance enhancements

#### Related Pull Requests
https://github.com/LabKey/platform/pull/1183
https://github.com/LabKey/premium/pull/190
https://github.com/LabKey/sampleManagement/pull/269

#### Changes
Put debug-only code under asserts
Lsid.toString() is expensive,  don't call in Lsid() constructor
Pattern.match() is expensive don't use in Lsid() constructor
PropertyService.getDomains(Set<String> domainKinds) is expensive, create new method getDomains(DomainKind dk)
Avoid creating new DefaultSchema (caching constructed TableInfo objects relies on this)
Avoid any expensive SQL generation work until we need to do it (ExprColumn, QuerySelect)
AssayWellExclusion service, use provider instead of protocol to avoid expensive protocol->provider lookup
Added EOF token to SQL grammar (forum recommendation)

### CODE REVIEW

File: api/src/org/labkey/api/exp/Lsid.java [hunk 1]
```diff
@@ -345,16 +359,16 @@ public LsidBuilder edit()
...
+            if (null == parts)
             {
                 valid = false;
                 return;
             }
-            authority = decodePart(m.group(1).toLowerCase());
-            namespace = decodePart(m.group(2));
-            objectId = decodePart(m.group(3));
-            version = decodePart(m.group(4));
+            authority = decodePart(parts[0]);
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/exp/Lsid.java]: We did a toLowerCase before.  Not needed?

File: api/src/org/labkey/api/exp/property/PropertyService.java [hunk 2]
```diff
@@ -66,9 +66,11 @@ static void setInstance(PropertyService impl)
     /** Get all the domains in the specified container and optionally project and shared. */
     List<? extends Domain> getDomains(Container container, User user, boolean includeProjectAndShared);
 
-    /** Get all the domains in the specified container and specified Domain Kinds. */
+    /** Get all the domains in the specified container and specified Domain Kinds. THIS IS SLOW */
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/exp/property/PropertyService.java]: Is the new method that uses a single DomainKind to be preferred here?  If so, perhaps worth a comment to that effect.

File: api/src/org/labkey/api/query/FilteredTable.java [hunk 3]
```diff
@@ -61,6 +62,8 @@
  */
 public class FilteredTable<SchemaType extends UserSchema> extends AbstractContainerFilterable implements ContainerFilterable
 {
+    final private static ButtonBarConfig BUTTONBAR_NOTSET = new ButtonBarConfig(new JSONObject());
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/query/FilteredTable.java]:
```suggestion
    final private static ButtonBarConfig BUTTON_BAR_NOT_SET = new ButtonBarConfig(new JSONObject());

```

File: api/src/org/labkey/api/query/FilteredTable.java [hunk 4]
```diff
@@ -714,4 +716,18 @@ public boolean hasPermission(@NotNull UserPrincipal user, @NotNull Class<? exten
         }
         return super.hasPermission(user, perm);
     }
+
+    @Override
+    public ButtonBarConfig getButtonBarConfig()
+    {
+        if (_buttonBarConfig == BUTTONBAR_NOTSET)
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/query/FilteredTable.java]: Comparison using this empty object is probably slightly less efficient than using an extra boolean `_isButtonBarSet` field but probably not enough to worry with.

File: assay/api-src/org/labkey/api/assay/AssayResultTable.java [hunk 5]
```diff
@@ -223,8 +223,8 @@ else if (baseColumn.isMvIndicatorColumn())
         AssayWellExclusionService svc = AssayWellExclusionService.getProvider(_protocol);
         if (svc != null)
         {
-            addColumn(svc.createExcludedColumn(this, _protocol));
-            addColumn(svc.createExclusionCommentColumn(this, _protocol));
+            addColumn(svc.createExcludedColumn(this, getUserSchema().getProvider()));
```
Comment (labkey-susanh) [inline assay/api-src/org/labkey/api/assay/AssayResultTable.java:226]: Is this change also wanted on line 223 or only for getting the exclusion columns that will be in the user schema?  Is this going to be a change in behavior?  Should we have tests to cover this?

File: assay/src/org/labkey/assay/TSVProtocolSchema.java [hunk 6]
```diff
@@ -122,9 +122,9 @@ private TableInfo createExclusionReportTable(ContainerFilter cf)
             AssayWellExclusionService svc = AssayWellExclusionService.getProvider(getProtocol());
             if (svc != null)
             {
-                var excludedByColumn = svc.createExcludedByColumn(this, getProtocol());
-                var excludedAtColumn = svc.createExcludedAtColumn(this, getProtocol());
-                var excludedCommentColumn = svc.createExclusionCommentColumn(this, getProtocol());
+                var excludedByColumn = svc.createExcludedByColumn(this, getProvider());
+                var excludedAtColumn = svc.createExcludedAtColumn(this, getProvider());
+                var excludedCommentColumn = svc.createExclusionCommentColumn(this, getProvider());
```
Comment (labkey-susanh) [inline assay/src/org/labkey/assay/TSVProtocolSchema.java:127]: Same question here.  Are we correcting behavior here so we might need test coverage?

File: experiment/src/org/labkey/experiment/api/property/PropertyServiceImpl.java [hunk 7]
```diff
@@ -204,6 +204,17 @@ public void registerDomainKind(DomainKind type)
         return getDomainsStream(container, user, domainKinds, includeProjectAndShared).collect(Collectors.toList());
     }
 
+    @Override
+    public List<? extends Domain> getDomains(Container container, User user, @NotNull DomainKind<?> dk, boolean includeProjectAndShared)
+    {
+        // Domain.getDomainKind() can be slow, instead just ask the passed in dk the domain matches
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/property/PropertyServiceImpl.java]:
```suggestion
        // Domain.getDomainKind() can be slow. Instead just ask the passed-in dk if the domain matches

```

File: query/src/org/labkey/query/sql/QField.java [hunk 8]
```diff
@@ -170,7 +171,15 @@ public int hashCode()
...
     public BaseColumnInfo createColumnInfo(SQLTableInfo table, String alias, Query query)
     {
-        ExprColumn ret = new ExprColumn(table, alias, getRelationColumn().getValueSql(), getRelationColumn().getJdbcType());
+        final QueryRelation.RelationColumn rcol = getRelationColumn();
+        ExprColumn ret = new ExprColumn(table, alias, null, getRelationColumn().getJdbcType())
+        {
+            @Override
+            public SQLFragment getValueSql(String tableAlias)
+            {
+                return rcol.getValueSql();  // delay call in case this causes unnecessary
```
Comment (labkey-susanh) [inline query/src/org/labkey/query/sql/QField.java]: unnecessary what?  Angst?

File: query/src/org/labkey/query/sql/QuerySelect.java [hunk 9]
```diff
@@ -2146,7 +2146,7 @@ void copyColumnAttributesTo(BaseColumnInfo to)
             {
                 if (_colinfo == null)
                 {
-                    _colinfo = expr.createColumnInfo(_sti, _aliasManager.decideAlias(getAlias()), _query);
+                    _colinfo = expr.createColumnInfo(_sti, "/*NOT AN OUTPUT COLUMN*/", _query);
```
Comment (labkey-susanh) [inline query/src/org/labkey/query/sql/QuerySelect.java:2150]: I don't pretend to know what's going on here, but this seems quite different than the previous version.  Is that intentional?  This second argument is supposed to be a name, but that seems to not be much like a decent column name.

File: query/src/org/labkey/query/sql/SqlParser.java [hunk 10]
```diff
@@ -1906,13 +1991,15 @@ public void testSql()
...
-					bad(sql);
+                    bad(sql);
                 }
                 catch (Throwable t)
                 {
                     fail(sql);
                 }
             }
+            long end = System.currentTimeMillis();
+//            System.out.println("SqlParser.testSql(): " + DateUtil.formatDuration(end-start));
```
Comment (labkey-susanh) [inline query/src/org/labkey/query/sql/SqlParser.java]: Remove commented out println?

**Review by labkey-susanh — COMMENTED** [review]
Just a few suggestions and some comments on what seem like changes in behavior.

File: assay/api-src/org/labkey/api/assay/AssayResultTable.java (hunk 5 above)
Comment (labkey-susanh) [reply assay/api-src/org/labkey/api/assay/AssayResultTable.java:226]: Ah, I see.  Signature changed.

File: assay/src/org/labkey/assay/TSVProtocolSchema.java (hunk 6 above)
Comment (labkey-susanh) [reply assay/src/org/labkey/assay/TSVProtocolSchema.java:127]: Never mind.  Signature changed.

**Review by labkey-susanh — COMMENTED** [review]
Looks like we've got a memory leak in some SM tests.

File: api/src/org/labkey/api/exp/Lsid.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/exp/Lsid.java]: unintentional difference.  Kinda weird that toLowerCase() is inside decodePart()

File: api/src/org/labkey/api/exp/property/PropertyService.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/exp/property/PropertyService.java]: added comment

File: api/src/org/labkey/api/query/FilteredTable.java (hunk 3 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/query/FilteredTable.java]: OK

File: experiment/src/org/labkey/experiment/api/property/PropertyServiceImpl.java (hunk 7 above)
Comment (labkey-matthewb) [reply experiment/src/org/labkey/experiment/api/property/PropertyServiceImpl.java]: OK

File: assay/src/org/labkey/assay/TSVProtocolSchema.java (hunk 6 above)
Comment (labkey-matthewb) [reply assay/src/org/labkey/assay/TSVProtocolSchema.java:127]: next time

File: query/src/org/labkey/query/sql/QField.java (hunk 8 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/QField.java]: check

File: query/src/org/labkey/query/sql/QuerySelect.java (hunk 9 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/QuerySelect.java:2150]: I'll add a comment

File: query/src/org/labkey/query/sql/SqlParser.java (hunk 10 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/sql/SqlParser.java]: _log.trace()

**Comment by labkey-matthewb** [comment]
> Looks like we've got a memory leak in some SM tests.

Got it already.

**Review by labkey-susanh — APPROVED** [review]
Changes here look good as long as the memory leak in SM goes away.

**Pull request merged**